

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Alignment &mdash; XPRESSpipe 0.2.4-beta documentation</title>
  

  
  
    <link rel="shortcut icon" href="../_static/xpressyourself.png"/>
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Read Quantification" href="count.html" />
    <link rel="prev" title="Read Pre-Processing" href="trim.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> XPRESSpipe
          

          
            
            <img src="../_static/xpressyourself.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">XPRESSpipe</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="beginner.html">Beginner’s Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="general-usage.html">General Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference-building.html">Curating References</a></li>
<li class="toctree-l1"><a class="reference internal" href="seRNAseq.html">Single-End RNA-seq Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="peRNAseq.html">Paired-End RNA-seq Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="riboseq.html">Ribosome Profiling Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="quality-control.html">Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis.html">Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="trim.html">Read Pre-Processing</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Alignment</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#single-end-rnaseq-alignment">Single-End RNAseq Alignment</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#arguments">Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#examples">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#paired-end-rnaseq-alignment">Paired-End RNAseq Alignment</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id1">Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id2">Examples</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="count.html">Read Quantification</a></li>
<li class="toctree-l1"><a class="reference internal" href="normalization.html">Normalize</a></li>
<li class="toctree-l1"><a class="reference internal" href="other-features.html">Other Features</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="updates.html">Updates</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">XPRESSpipe</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Alignment</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/content/align.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="alignment">
<h1>Alignment<a class="headerlink" href="#alignment" title="Permalink to this headline">¶</a></h1>
<div class="line-block">
<div class="line">In order to quantify transcription on a transcript to transcript basis, individual reads called during sequencing must be mapped to the genome. While there are multiple alignment software packages available, XPRESSpipe uses a current version of <a class="reference external" href="https://github.com/alexdobin/STAR">STAR</a> to perform this step in transcription quantification for several reasons:</div>
<div class="line">- Performance: While computationally greedy (a human genome alignment requires upwards of 30 Gb RAM), the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5792058/">performance and accuracy is superior to the majority of other splice aware aligners currently available</a></div>
<div class="line">- Splice Junction Aware: STAR is capable of mapping reads spanning a splice junction, where more traditional packages, such as Bowtie, are incapable of doing so and are better suited for tasks such as genome alignment.</div>
<div class="line">- Standard: The foundation of the pipeline used in XPRESSpipe is based in the <a class="reference external" href="https://docs.gdc.cancer.gov/Data/Bioinformatics_Pipelines/Expression_mRNA_Pipeline/">TCGA</a> standards for RNAseq alignment. This method utilizes a guided or 2-pass alignment program. In the guided alignment, a GTF with annotated splice junctions is used to guide the alignments over splice juntions. In the 2-pass alignment, reads are mapped across the genome to identify novel splice junctions. These new annotations are then incorporated into the reference index and reads are re-aligned with this new reference. While more time-intensive, this step can aid in aligning across these junctions, especially in organisms where the transcriptome is not as well annotated.</div>
<div class="line">- Variant Aware: The user can provide a VCF, such as those provided by <a class="reference external" href="ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606/VCF/">ClinVar</a> and <a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar/">dbSNP</a>. These files are useful in integrating information about common or disease nucleotide variants into the RNA-Seq alignment stage. The files you use should match the build of the genome you are using (i.e., if using Homo Sapiens GRCh38, these builds should match between curated reference files and VCF file).</div>
</div>
<div class="section" id="single-end-rnaseq-alignment">
<h2>Single-End RNAseq Alignment<a class="headerlink" href="#single-end-rnaseq-alignment" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">The following runs single-end reads alignment using the specified XPRESSpipe-formatted reference directory.</div>
<div class="line">Notes:</div>
<div class="line">- For the <code class="xref py py-data docutils literal notranslate"><span class="pre">--sjdbOverhang</span></code> argument, the same value should be entered that was used when creating the STAR reference files.</div>
<div class="line">- Ribosome profiling data can be run as a single-end RNA-seq</div>
</div>
<div class="section" id="arguments">
<h3>Arguments<a class="headerlink" href="#arguments" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line">The help menu can be accessed by calling the following from the command line:</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe align --help
</pre></div>
</div>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Required Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-i</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--input</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to input directory</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-o</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--output</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to output directory</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-r</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--reference</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to parent organism reference directory (must have a file called transcripts.gtf within)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-t</span> <span class="pre">&lt;SE</span> <span class="pre">or</span> <span class="pre">PE&gt;,</span> <span class="pre">--type</span> <span class="pre">&lt;SE</span> <span class="pre">or</span> <span class="pre">PE&gt;</span></code></p></td>
<td><p>Sequencing type (“SE” for single-end, “PE” for paired-end)</p></td>
</tr>
</tbody>
</table>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Optional Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--two-pass</span></code></p></td>
<td><p>Use a two-pass STAR alignment for novel splice junction discovery</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--no_multimappers&gt;</span></code></p></td>
<td><p>Include flag to remove multimapping reads to be output and used in downstream analyses</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--deduplicate</span></code></p></td>
<td><p>Include flag to quantify reads with de-duplication (will search for files with suffix <code class="xref py py-data docutils literal notranslate"><span class="pre">_dedupRemoved.bam</span></code>)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--vcf</span> <span class="pre">&lt;/path/to/file.vcf&gt;</span></code></p></td>
<td><p>Provide full path and file name to VCF file if you would like detect personal variants overlapping alignments</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--output_bed</span></code></p></td>
<td><p>Include flag to output BED files for each aligned file</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--sjdbOverhang</span> <span class="pre">&lt;sjdbOverhang_amount&gt;</span></code></p></td>
<td><p>Specify length of genomic sequences for constructing splice-aware reference. Ideal length is <code class="xref py py-data docutils literal notranslate"><span class="pre">read</span> <span class="pre">length</span> <span class="pre">-</span> <span class="pre">1</span></code>, so for 2x100bp paired-end reads, you would use 100 - 1 = 99. However, the default value of <code class="xref py py-data docutils literal notranslate"><span class="pre">100</span></code> should work in most cases</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--mismatchRatio</span> <span class="pre">&lt;mismatchRatio&gt;</span></code></p></td>
<td><p>Alignment ratio of mismatches to mapped length is less than this value. See STAR documentation for more information on setting this parameter</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--seedSearchStartLmax</span> <span class="pre">&lt;seedSearchStartLmax&gt;</span></code></p></td>
<td><p>Adjusting this parameter by providing a lower number will improve mapping sensitivity (recommended value = 15 for reads ~ 25 nts). See STAR documentation for more information on setting this parameter</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">genome_size</span></code></p></td>
<td><p>Only needs to be changed if this argument was provided curing reference building AND using a two-pass alignment. Enter the size of your organism’s genome in nucleotides</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-m</span></code></p></td>
<td><p>Number of max processors to use for tasks (default: No limit)</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="examples">
<h3>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line"><strong>Example 1 – Single-end RNAseq alignment:</strong></div>
<div class="line">- Raw reads are <code class="xref py py-data docutils literal notranslate"><span class="pre">fastq</span></code>-like and found in the <code class="xref py py-data docutils literal notranslate"><span class="pre">-i</span> <span class="pre">/path/to/input/files/</span></code> directory. Can be uncompressed or compressed via <code class="xref py py-data docutils literal notranslate"><span class="pre">gz</span></code> or <code class="xref py py-data docutils literal notranslate"><span class="pre">zip</span></code></div>
<div class="line">- A general output directory has been created, <code class="xref py py-data docutils literal notranslate"><span class="pre">-o</span> <span class="pre">riboseq_out/</span></code></div>
<div class="line">- <code class="xref py py-data docutils literal notranslate"><span class="pre">--type</span></code> is specified as ‘SE’ and path to parent reference directory is provided</div>
<div class="line">- The value for <code class="xref py py-data docutils literal notranslate"><span class="pre">--sjdbOverhang</span></code> used in reference creation is provided. Failure to do so will trigger an error</div>
<div class="line">- BED and BIGWIG files will be output in their own directories in <code class="xref py py-data docutils literal notranslate"><span class="pre">output</span></code></div>
<div class="line">- All other arguments use the default value</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe align -i /path/to/input/files/ -o riboseq_out/ -t SE -r /path/to/reference/ --sjdbOverhang <span class="m">49</span> --output_bed --output_bigwig
</pre></div>
</div>
</div>
</div>
<div class="section" id="paired-end-rnaseq-alignment">
<h2>Paired-End RNAseq Alignment<a class="headerlink" href="#paired-end-rnaseq-alignment" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">The following runs paired-end reads alignment using the specified XPRESSpipe-formatted reference directory.</div>
<div class="line">Notes:</div>
<div class="line">- For the <code class="xref py py-data docutils literal notranslate"><span class="pre">--sjdbOverhang</span></code> argument, the same value should be entered that was used when creating the STAR reference files.</div>
</div>
<div class="section" id="id1">
<h3>Arguments<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line">The help menu can be accessed by calling the following from the command line:</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe align --help
</pre></div>
</div>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Required Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-i</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--input</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to input directory</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-o</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--output</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to output directory</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-r</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--reference</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to parent organism reference directory</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-t</span> <span class="pre">&lt;SE</span> <span class="pre">or</span> <span class="pre">PE&gt;,</span> <span class="pre">--type</span> <span class="pre">&lt;SE</span> <span class="pre">or</span> <span class="pre">PE&gt;</span></code></p></td>
<td><p>Sequencing type (“SE” for single-end, “PE” for paired-end)</p></td>
</tr>
</tbody>
</table>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Optional Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--output_bed</span></code></p></td>
<td><p>Include flag to output BED files for each aligned file</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--output_bigwig</span></code></p></td>
<td><p>Include flag to output bigwig files for each aligned file</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--sjdbOverhang</span> <span class="pre">&lt;sjdbOverhang_amount&gt;</span></code></p></td>
<td><p>Specify length of genomic sequences for constructing splice-aware reference. Ideal length is <code class="xref py py-data docutils literal notranslate"><span class="pre">read</span> <span class="pre">length</span> <span class="pre">-</span> <span class="pre">1</span></code>, so for 2x100bp paired-end reads, you would use 100 - 1 = 99. However, the default value of <code class="xref py py-data docutils literal notranslate"><span class="pre">100</span></code> should work in most cases</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-m</span></code></p></td>
<td><p>Number of max processors to use for tasks (default: No limit)</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id2">
<h3>Examples<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line"><strong>Example 1 – Paired-end RNAseq alignment:</strong></div>
<div class="line">- Raw reads are <code class="xref py py-data docutils literal notranslate"><span class="pre">fastq</span></code>-like and found in the <code class="xref py py-data docutils literal notranslate"><span class="pre">-i</span> <span class="pre">pe_test/</span></code> directory. Can be uncompressed or compressed via <code class="xref py py-data docutils literal notranslate"><span class="pre">gz</span></code> or <code class="xref py py-data docutils literal notranslate"><span class="pre">zip</span></code></div>
<div class="line">- A general output directory has been created, <code class="xref py py-data docutils literal notranslate"><span class="pre">-o</span> <span class="pre">pe_out/</span></code></div>
<div class="line">- <code class="xref py py-data docutils literal notranslate"><span class="pre">--type</span></code> is specified as ‘PE’ and path to parent reference directory is provided</div>
<div class="line">- The value for <code class="xref py py-data docutils literal notranslate"><span class="pre">--sjdbOverhang</span></code> used in reference creation is provided. Failure to do so will trigger an error. In this case, since the reference was created using default values, the optional flag is not used</div>
<div class="line">- BED and BIGWIG files are not output</div>
<div class="line">- All other arguments use the default value</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe align -i /path/to/input/files/ -o riboseq_out -t PE -r /path/to/reference/
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="count.html" class="btn btn-neutral float-right" title="Read Quantification" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="trim.html" class="btn btn-neutral float-left" title="Read Pre-Processing" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Jordan A. Berg

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>