

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Analysis &mdash; XPRESSpipe 0.2.4-beta documentation</title>
  

  
  
    <link rel="shortcut icon" href="../_static/xpressyourself.png"/>
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Read Pre-Processing" href="trim.html" />
    <link rel="prev" title="Quality Control" href="quality-control.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> XPRESSpipe
          

          
            
            <img src="../_static/xpressyourself.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">XPRESSpipe</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="beginner.html">Beginner’s Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="general-usage.html">General Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference-building.html">Curating References</a></li>
<li class="toctree-l1"><a class="reference internal" href="seRNAseq.html">Single-End RNA-seq Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="peRNAseq.html">Paired-End RNA-seq Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="riboseq.html">Ribosome Profiling Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="quality-control.html">Quality Control</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Analysis</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#differential-expression-analysis">Differential Expression Analysis</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#arguments">Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#examples">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#rrna-probe">rRNA Probe</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id2">Examples</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="trim.html">Read Pre-Processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="align.html">Alignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="count.html">Read Quantification</a></li>
<li class="toctree-l1"><a class="reference internal" href="normalization.html">Normalize</a></li>
<li class="toctree-l1"><a class="reference internal" href="other-features.html">Other Features</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="updates.html">Updates</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">XPRESSpipe</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Analysis</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/content/analysis.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="analysis">
<span id="analysis-link"></span><h1>Analysis<a class="headerlink" href="#analysis" title="Permalink to this headline">¶</a></h1>
<div class="section" id="differential-expression-analysis">
<h2>Differential Expression Analysis<a class="headerlink" href="#differential-expression-analysis" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">Differential Expression analysis allows one to determine significantly enriched or depleted genes between two conditions. XPESSpipe acts as a wrapper for <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4302049/">DESeq2</a>. Please refer to its <a class="reference external" href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">documentation</a> for more information.</div>
<div class="line">NOTE: If intending to use the <code class="xref py py-data docutils literal notranslate"><span class="pre">diffxpress</span></code> sub-module, you need to have used <code class="xref py py-data docutils literal notranslate"><span class="pre">--quantification_method</span> <span class="pre">htseq</span></code> for during read quantification and <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4302049/">DESeq2</a> requires integer count data.</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
<div class="line">Assumptions:</div>
<div class="line-block">
<div class="line">- R is installed on your machine and is in your $PATH (this should be handled in the installation)</div>
<div class="line">- All input files are tab-delimited (with .txt or .tsv suffix)</div>
<div class="line">- Design formula does not include the tilde (~) and there are no spaces</div>
<div class="line">- Your base parameter in a given column in the sample_info file must be first alphabetically. If this is not the base, you can append letters to the beginning to force alphabetical order</div>
</div>
</div>
<div class="section" id="arguments">
<h3>Arguments<a class="headerlink" href="#arguments" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line">The help menu can be accessed by calling the following from the command line:</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe diffxpress --help
</pre></div>
</div>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Required Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-i</span> <span class="pre">&lt;path/filename.tsv&gt;</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--input</span> <span class="pre">&lt;path/filename.tsv&gt;</span></code></p></td>
<td><p>Path and file name of expression counts matrix</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-s</span> <span class="pre">&lt;path/filename.tsv&gt;</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--sample</span> <span class="pre">&lt;path/filename.tsv&gt;</span></code></p></td>
<td><p>Path and file name of sample information matrix</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--design</span> <span class="pre">&lt;formula&gt;</span></code></p></td>
<td><p>Design formula for differential expression analysis (spaces in command line are conserved in input string. DO NOT INCLUDE ~ OR SPACES IN FORMULA IN COMMAND LINE, will be automatically added)</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="examples">
<h3>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line"><strong>Example 1 – Analyze RNA-seq data</strong></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;/path/to/expression_counts.tsv&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">data</span>
                <span class="n">s1</span>  <span class="n">s2</span>  <span class="n">s3</span>  <span class="n">s4</span>  <span class="o">...</span>
<span class="n">ENSG00000227232</span> <span class="mi">66</span>  <span class="mi">59</span>  <span class="mi">1</span>   <span class="mi">82</span>  <span class="o">...</span>
<span class="n">ENSG00000240361</span> <span class="mi">35</span>  <span class="mi">0</span>   <span class="mi">7</span>   <span class="mi">72</span>  <span class="o">...</span>
<span class="n">ENSG00000238009</span> <span class="mi">20</span>  <span class="mi">70</span>  <span class="mi">85</span>  <span class="mi">78</span>  <span class="o">...</span>
<span class="n">ENSG00000241860</span> <span class="mi">96</span>  <span class="mi">7</span>   <span class="mi">93</span>  <span class="mi">38</span>  <span class="o">...</span>
<span class="n">ENSG00000187634</span> <span class="mi">73</span>  <span class="mi">41</span>  <span class="mi">92</span>  <span class="mi">77</span>  <span class="o">...</span>
<span class="o">&gt;</span> <span class="n">batch</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;/path/to/sample_info.tsv&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">batch</span>
  <span class="n">Sample</span>  <span class="n">Replicate</span> <span class="n">Condition</span>
<span class="mi">0</span> <span class="n">s1</span>      <span class="n">rep1</span>      <span class="n">a_WT</span>
<span class="mi">1</span> <span class="n">s2</span>      <span class="n">rep2</span>      <span class="n">a_WT</span>
<span class="mi">2</span> <span class="n">s3</span>      <span class="n">rep1</span>      <span class="n">a_WT</span>
<span class="mi">3</span> <span class="n">s4</span>      <span class="n">rep2</span>      <span class="n">a_WT</span>
<span class="mi">4</span> <span class="n">s5</span>      <span class="n">rep1</span>      <span class="n">b_EXP</span>
<span class="mi">5</span> <span class="n">s6</span>      <span class="n">rep2</span>      <span class="n">b_EXP</span>
<span class="mi">6</span> <span class="n">s7</span>      <span class="n">rep1</span>      <span class="n">b_EXP</span>
<span class="mi">7</span> <span class="n">s8</span>      <span class="n">rep2</span>      <span class="n">b_EXP</span>
</pre></div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe diffxpress -i test_r/test_dataset.tsv --sample test_r/sample_info.tsv --design Condition
</pre></div>
</div>
<div class="line-block">
<div class="line"><strong>Example 2 – Analyze RNA-seq data that was prepared in different batches:</strong></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;/path/to/expression_counts.tsv&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">data</span>
                <span class="n">s1</span>  <span class="n">s2</span>  <span class="n">s3</span>  <span class="n">s4</span>  <span class="o">...</span>
<span class="n">ENSG00000227232</span> <span class="mi">66</span>  <span class="mi">59</span>  <span class="mi">1</span>   <span class="mi">82</span>  <span class="o">...</span>
<span class="n">ENSG00000240361</span> <span class="mi">35</span>  <span class="mi">0</span>   <span class="mi">7</span>   <span class="mi">72</span>  <span class="o">...</span>
<span class="n">ENSG00000238009</span> <span class="mi">20</span>  <span class="mi">70</span>  <span class="mi">85</span>  <span class="mi">78</span>  <span class="o">...</span>
<span class="n">ENSG00000241860</span> <span class="mi">96</span>  <span class="mi">7</span>   <span class="mi">93</span>  <span class="mi">38</span>  <span class="o">...</span>
<span class="n">ENSG00000187634</span> <span class="mi">73</span>  <span class="mi">41</span>  <span class="mi">92</span>  <span class="mi">77</span>  <span class="o">...</span>
<span class="o">&gt;</span> <span class="n">batch</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;/path/to/sample_info.tsv&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">batch</span>
  <span class="n">Sample</span>  <span class="n">Replicate</span> <span class="n">Condition</span> <span class="n">Batch</span>
<span class="mi">0</span> <span class="n">s1</span>      <span class="n">rep1</span>      <span class="n">a_WT</span>      <span class="n">batch1</span>
<span class="mi">1</span> <span class="n">s2</span>      <span class="n">rep2</span>      <span class="n">a_WT</span>      <span class="n">batch1</span>
<span class="mi">2</span> <span class="n">s3</span>      <span class="n">rep1</span>      <span class="n">a_WT</span>      <span class="n">batch1</span>
<span class="mi">3</span> <span class="n">s4</span>      <span class="n">rep2</span>      <span class="n">a_WT</span>      <span class="n">batch1</span>
<span class="mi">4</span> <span class="n">s5</span>      <span class="n">rep1</span>      <span class="n">b_EXP</span>     <span class="n">batch2</span>
<span class="mi">5</span> <span class="n">s6</span>      <span class="n">rep2</span>      <span class="n">b_EXP</span>     <span class="n">batch2</span>
<span class="mi">6</span> <span class="n">s7</span>      <span class="n">rep1</span>      <span class="n">b_EXP</span>     <span class="n">batch2</span>
<span class="mi">7</span> <span class="n">s8</span>      <span class="n">rep2</span>      <span class="n">b_EXP</span>     <span class="n">batch2</span>
</pre></div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe diffxpress -i test_r/test_dataset.tsv --sample test_r/sample_info.tsv --design Condition+Batch
</pre></div>
</div>
<div class="line-block">
<div class="line"><strong>Example 3 – Analyze ribosome profiling data:</strong></div>
<div class="line">For ribosome profiling, you need to divide the footprint samples by their corresponding mRNA sample to account for translation efficiency</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;/path/to/expression_counts.tsv&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">data</span>
                <span class="n">s1_fp</span>   <span class="n">s1_rna</span>  <span class="n">s2_fp</span>   <span class="n">s2_rna</span>  <span class="o">...</span>
<span class="n">ENSG00000227232</span> <span class="mi">66</span>      <span class="mi">59</span>      <span class="mi">1</span>       <span class="mi">82</span>      <span class="o">...</span>
<span class="n">ENSG00000240361</span> <span class="mi">35</span>      <span class="mi">0</span>       <span class="mi">7</span>       <span class="mi">72</span>      <span class="o">...</span>
<span class="n">ENSG00000238009</span> <span class="mi">20</span>      <span class="mi">70</span>      <span class="mi">85</span>      <span class="mi">78</span>      <span class="o">...</span>
<span class="n">ENSG00000241860</span> <span class="mi">96</span>      <span class="mi">7</span>       <span class="mi">93</span>      <span class="mi">38</span>      <span class="o">...</span>
<span class="n">ENSG00000187634</span> <span class="mi">73</span>      <span class="mi">41</span>      <span class="mi">92</span>      <span class="mi">77</span>      <span class="o">...</span>
<span class="o">&gt;</span> <span class="n">batch</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;/path/to/sample_info.tsv&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">batch</span>
  <span class="n">Sample</span>  <span class="n">Replicate</span> <span class="n">Condition</span> <span class="n">Type</span>
<span class="mi">0</span> <span class="n">s1_fp</span>   <span class="n">rep1</span>      <span class="n">a_WT</span>      <span class="n">RPF</span>
<span class="mi">1</span> <span class="n">s1_rna</span>  <span class="n">rep1</span>      <span class="n">a_WT</span>      <span class="n">RNA</span>
<span class="mi">2</span> <span class="n">s2_fp</span>   <span class="n">rep2</span>      <span class="n">a_WT</span>      <span class="n">RPF</span>
<span class="mi">3</span> <span class="n">s2_rna</span>  <span class="n">rep2</span>      <span class="n">a_WT</span>      <span class="n">RNA</span>
<span class="mi">4</span> <span class="n">s3_fp</span>   <span class="n">rep1</span>      <span class="n">b_EXP</span>     <span class="n">RPF</span>
<span class="mi">5</span> <span class="n">s3_rna</span>  <span class="n">rep1</span>      <span class="n">b_EXP</span>     <span class="n">RNA</span>
<span class="mi">6</span> <span class="n">s4_fp</span>   <span class="n">rep2</span>      <span class="n">b_EXP</span>     <span class="n">RPF</span>
<span class="mi">7</span> <span class="n">s4_rna</span>  <span class="n">rep2</span>      <span class="n">b_EXP</span>     <span class="n">RNA</span>
</pre></div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe diffxpress -i test_r/test_dataset.tsv --sample test_r/sample_info.tsv --design Type+Condition+Type:Condition
</pre></div>
</div>
</div>
</div>
<div class="section" id="rrna-probe">
<h2>rRNA Probe<a class="headerlink" href="#rrna-probe" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">Ribosome RNA (rRNA) contamination is common in RNA-seq library preparation. As the bulk of RNA in a cell at any given time is dedicated to rRNA, and as these rRNA sequences are relatively few and therefore highly repeated, depletion of these sequences is often desired in order to have better depth of coverage of non-rRNA sequences. In order to facilitate this depletion, many commercial kits are available that target specific rRNA sequences for depletion, or that enrich mRNA polyA tails. However, and especially in the case of ribosome profiling experiments, where RNA is digested to create ribosome footprints that commercial depletion kits won’t detect and polyA selection kits are inoperable as footprints will not have the requisite polyA sequence. To this end, <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/28579404">custom rRNA probes</a> are recommended, and the <code class="xref py py-data docutils literal notranslate"><span class="pre">rrnaProbe</span></code> sub-module was designed to facilitate this process.</div>
<div class="line"><code class="xref py py-data docutils literal notranslate"><span class="pre">rrnaProbe</span></code> works by doing the following:</div>
<div class="line">1. Run FASTQC to detect over-represented sequences</div>
<div class="line">2. Collate these sequences to determine consensus fragments</div>
<div class="line">3. Output rank ordered list of over-represented fragments within the appropriate length range to target for depletion</div>
<div class="line">NOTE: BLAST capability to verify over-represented consensus fragments are indeed rRNA sequences is not yet incorporated, so any sequences that will be used as probes should be BLAST-verified first.</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe rrnaProbe --help
</pre></div>
</div>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Required Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-i</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--input</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to zipped FASTQC files</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-o</span> <span class="pre">&lt;/path/filename&gt;,</span> <span class="pre">--output</span> <span class="pre">&lt;/path/filename&gt;</span></code></p></td>
<td><p>Path and file name to write output</p></td>
</tr>
</tbody>
</table>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Optional Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-m</span> <span class="pre">&lt;value&gt;,</span> <span class="pre">--min_overlap</span> <span class="pre">&lt;value&gt;</span></code></p></td>
<td><p>Minimum number of bases that must match on a side to combine sequences (default: 5)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--footprint_only</span></code></p></td>
<td><p>Only take zip files that are ribosome profiling footprints (file names must contain “FP”, “RPF”, or “FOOTPRINT”)</p></td>
</tr>
</tbody>
</table>
<div class="section" id="id2">
<h3>Examples<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line"><strong>Example 1 – Generate rank-ordered list of over-represented sequences</strong></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>$ xpresspipe rrnaProbe -i riboprof_out/fastqc_out/ -o riboprof_out/sequences.txt --footprint_only

TTGATGATTCATAATAACTTTTCGAATCGCAT    514832
TATAAATCATTTGTATACGACTTAGAT         121739
TTGATGATTCATAATAACTTTTCGAATCGCAT    15776
TTTGATGATTCATAATAACTTTTCGAATCGCAC   33325
ATAAATCATTTGTATACGACTTAGAC          13603
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="trim.html" class="btn btn-neutral float-right" title="Read Pre-Processing" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="quality-control.html" class="btn btn-neutral float-left" title="Quality Control" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Jordan A. Berg

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>