

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Beginner’s Guide &mdash; XPRESSpipe 0.2.4-beta documentation</title>
  

  
  
    <link rel="shortcut icon" href="../_static/xpressyourself.png"/>
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Installation" href="installation.html" />
    <link rel="prev" title="Quickstart" href="quickstart.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> XPRESSpipe
          

          
            
            <img src="../_static/xpressyourself.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">XPRESSpipe</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Beginner’s Guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#first-steps">First Steps</a></li>
<li class="toctree-l2"><a class="reference internal" href="#install-xpresspipe">Install XPRESSpipe</a></li>
<li class="toctree-l2"><a class="reference internal" href="#generate-reference-files">Generate Reference Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#process-raw-sequencing-files">Process Raw Sequencing Files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sequencing-metrics">Sequencing Metrics</a></li>
<li class="toctree-l3"><a class="reference internal" href="#library-complexity">Library Complexity</a></li>
<li class="toctree-l3"><a class="reference internal" href="#metagene-analysis">Metagene Analysis</a></li>
<li class="toctree-l3"><a class="reference internal" href="#periodicity-ribosome-profiling">Periodicity (Ribosome Profiling)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#count-data-and-downstream-analysis">Count Data and Downstream Analysis</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#supercomputing">Supercomputing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#install">Install</a></li>
<li class="toctree-l3"><a class="reference internal" href="#run-data">Run Data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#explore-the-data">Explore the Data</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="general-usage.html">General Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference-building.html">Curating References</a></li>
<li class="toctree-l1"><a class="reference internal" href="seRNAseq.html">Single-End RNA-seq Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="peRNAseq.html">Paired-End RNA-seq Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="riboseq.html">Ribosome Profiling Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="quality-control.html">Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis.html">Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="trim.html">Read Pre-Processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="align.html">Alignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="count.html">Read Quantification</a></li>
<li class="toctree-l1"><a class="reference internal" href="normalization.html">Normalize</a></li>
<li class="toctree-l1"><a class="reference internal" href="other-features.html">Other Features</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="updates.html">Updates</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">XPRESSpipe</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Beginner’s Guide</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/content/beginner.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="beginner-s-guide">
<span id="beginners-link"></span><h1>Beginner’s Guide<a class="headerlink" href="#beginner-s-guide" title="Permalink to this headline">¶</a></h1>
<div class="section" id="first-steps">
<h2>First Steps<a class="headerlink" href="#first-steps" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">If this is your first time doing any programming, congratulations! You are embarking upon a very rewarding path. As with learning any new spoken language, there is a learning curve associated with learning a computer language. While XPRESSpipe is aimed at reducing the majority, if not (hopefully) all of the overhead associated with processing this data, using this software will still require some effort, just as would learning any new language or laboratory technique.</div>
</div>
<div class="line-block">
<div class="line">XPRESSpipe is run through the <a class="reference external" href="https://en.wikipedia.org/wiki/Command-line_interface">command line interface</a> (or <a class="reference external" href="https://www.youtube.com/watch?v=kqUR3KtWbTk">CLI</a>). This may seem daunting, but luckily, several free online courses are available to quickly catch you up to speed on some of the basics that will be required to use this software. We recommend Codecademy’s CLI course, which you can find <a class="reference external" href="https://www.codecademy.com/learn/learn-the-command-line">here</a> and should take only a couple of hours (Codecademy estimates ~10 hours, but you probably don’t need to finish the course to be ready to use XPRESSpipe. The purpose of this is to help you become more comfortable with the command line). We recommend watching the walkthrough videos found on the <a class="reference internal" href="quickstart.html"><span class="doc">quickstart</span></a> page.</div>
</div>
<div class="line-block">
<div class="line">Once you’re ready to jump into the command line, we can get rolling! For the steps below, we’re going to assume we are on an Mac operating system and provide examples under this pretext, but this software is compatible with any Linux-like operating system and the syntax is largely the same (sorry Windows users! – but if you have a newer version of Windows, you may be able to use a Linux-flavored environment).</div>
</div>
</div>
<div class="section" id="install-xpresspipe">
<h2>Install XPRESSpipe<a class="headerlink" href="#install-xpresspipe" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">- Please refer to the <a class="reference internal" href="installation.html"><span class="doc">installation documentation</span></a> or the walkthrough video below:</div>
</div>
<embed>
    <script id="asciicast-256347" src="https://asciinema.org/a/256347.js" async></script>
</embed></div>
<div class="section" id="generate-reference-files">
<h2>Generate Reference Files<a class="headerlink" href="#generate-reference-files" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">- Before we can process our raw RNA-seq data, we need to create a reference directory (or for a folder, in other terms). In this example, we will be working with human-derived RNA-seq data, so let’s perform the following in the command line:</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ <span class="nb">cd</span> ~/Desktop
$ mkdir reference_folder
$ mkdir reference_folder/fasta_files
</pre></div>
</div>
<div class="line-block">
<div class="line">1. The first command helped us navigate to the Desktop. The “~” icon is a shortcut for the User directory, and every directory needs to be separated by a “/”</div>
<div class="line">2. The second command created a new folder in the Desktop directory called <code class="xref py py-data docutils literal notranslate"><span class="pre">reference_folder</span></code></div>
<div class="line">3. The third command created a new folder in the reference directory for intermediate reference files</div>
</div>
<div class="line-block">
<div class="line">- Now let’s get the reference files. We’re going to do this directly in the command line, but if you have trouble with this, I will explain an alternative afterwards. Quick note, because the next lines of code are a bit long, I used the “” character to indicate I am continuing the command in the next line. You not include these characters when executing the command, they just help make the code more readable. We will first read the retrieval commands into a file which will additionally act as a log file for the version for the genome version we are using.</div>
<div class="line">- You should modify the the variable calls between the <code class="xref py py-data docutils literal notranslate"><span class="pre">#</span></code> signs. For <code class="xref py py-data docutils literal notranslate"><span class="pre">GTF_URL</span></code>, you should change the URL currently provided to the one appropriate for your organism of interest. Make sure you are downloading the GTF file and NOT the GFF file. For <code class="xref py py-data docutils literal notranslate"><span class="pre">FASTA_URL</span></code>, you should do the same as before with the URL to the chromosome DNA FASTA files, but you should only copy the URL up to “chromosome”, but not include the chromosome identifier. For <code class="xref py py-data docutils literal notranslate"><span class="pre">CHROMOSOMES</span></code>, type out the chromosome identifiers you want to download between the ” characters with a space between each.</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ <span class="nb">cd</span> reference_folder/

<span class="c1">### Change these ###</span>
$ <span class="nb">echo</span> <span class="s1">&#39;GTF_URL=ftp://ftp.ensembl.org/pub/release-97/gtf/homo_sapiens/Homo_sapiens.GRCh38.97.gtf.gz&#39;</span> &gt;&gt; fetch.sh
$ <span class="nb">echo</span> <span class="s1">&#39;FASTA_URL=ftp://ftp.ensembl.org/pub/release-97/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.chromosome&#39;</span> &gt;&gt; fetch.sh
$ <span class="nb">echo</span> <span class="s1">&#39;CHROMOSOMES=&quot;1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 X Y&quot;&#39;</span>
<span class="c1">####################</span>

$ <span class="nb">echo</span> <span class="s1">&#39;curl -O $GTF_URL&#39;</span> &gt;&gt; fetch.sh
$ <span class="nb">echo</span> <span class="s1">&#39;gzip -d Homo_sapiens.GRCh38.97.gtf.gz&#39;</span> &gt;&gt; fetch.sh
$ <span class="nb">echo</span> <span class="s1">&#39;mv Homo_sapiens.GRCh38.97.gtf transcripts.gtf&#39;</span> &gt;&gt; fetch.sh
$ <span class="nb">echo</span> <span class="s1">&#39;cd fasta_files/&#39;</span> &gt;&gt; fetch.sh
$ <span class="nb">echo</span> <span class="s1">&#39;for X in $CHROMOSOMES; &#39;</span> &gt;&gt; fetch.sh
$ <span class="nb">echo</span> <span class="s1">&#39;do curl -O ftp://ftp.ensembl.org/pub/release-97/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.chromosome.${X}.fa.gz; done &#39;</span> &gt;&gt; fetch.sh
$ <span class="nb">echo</span> <span class="s1">&#39;gzip -d *.gz&#39;</span> &gt;&gt; fetch.sh
$ <span class="nb">echo</span> <span class="s1">&#39;cd ../&#39;</span> &gt;&gt; fetch.sh
$ bash fetch.sh
</pre></div>
</div>
<div class="line-block">
<div class="line">- Let’s discuss what we just did:</div>
<div class="line">1. We navigated into the reference folder, downloaded a GTF reference file and unzipped it, then navigated to the <code class="xref py py-data docutils literal notranslate"><span class="pre">fasta_file</span></code> directory to download the raw reference data and unzipped it. Finally, we returned to the main reference directory.</div>
<div class="line">2. If this didn’t work, we can navigate to <a class="reference external" href="https://www.ensembl.org/">Ensembl</a> to download the relevant data. We need to get the GTF and DNA chromosomal FASTA files for our organism of interest. The link to the chromosome sequence files actually contains more files than we need. We just need the files that start with <code class="xref py py-data docutils literal notranslate"><span class="pre">Homo_sapiens.GRCh38.dna.chromosome</span></code>. You can download them, move them to the appropriate directories within your reference directory, and unzip the files by double-clicking on them.</div>
</div>
<div class="line-block">
<div class="line">- Now we need to curate these references files into something the sequencing alignment software can use. Since we are using ribosome profiling data, we want a reference that will allow us to <a class="reference external" href="https://www.cell.com/cms/10.1016/j.celrep.2016.01.043/attachment/257faf34-ff8f-4071-a642-bfdb531c75b8/mmc1">avoid mapping to the 5’ and 3’ ends of genes</a>. We also don’t want to align to anything but protein coding genes. Finally, we want to quantify to the longest transcript. This last bit just helps the software avoid confusion when a gene has multiple splice variants to choose from. Since this is short read sequencing (let’s say we were doing 50 bp single-end sequencing), we also want to factor this into the curation of the reference (see the <code class="xref py py-data docutils literal notranslate"><span class="pre">--sjdbOverhang</span></code> argument below).</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe curateReference <span class="se">\</span>
              --output ./ <span class="se">\</span>
              --fasta fasta_files/ <span class="se">\</span>
              --gtf ./transcripts.gtf <span class="se">\</span>
              --protein_coding <span class="se">\</span>
              --truncate <span class="se">\</span>
              --sjdbOverhang <span class="m">49</span>

<span class="c1">### or ###</span>

$ xpresspipe build

<span class="c1">### And then choose the curate option ###</span>
</pre></div>
</div>
<div class="line-block">
<div class="line">- The truncation option is only necessary when using XPRESSpipe to process ribosome profiling samples and their associated RNA-seq samples.</div>
<div class="line">- If interested in quantifying miRNA, etc, leave out the <code class="xref py py-data docutils literal notranslate"><span class="pre">--protein_coding</span></code> argument.</div>
<div class="line">- If running sequencing where the read (single-end) or mates not equal to 100 bp, you will want to change the <code class="xref py py-data docutils literal notranslate"><span class="pre">--sjdbOverhang</span></code> argument to be the length of one of the paired-end reads - 1, so if we ran 2x100bp sequencing, we would specify <code class="xref py py-data docutils literal notranslate"><span class="pre">--sjdbOverhang</span> <span class="pre">99</span></code> (although in this case, the default of <code class="xref py py-data docutils literal notranslate"><span class="pre">--sjdbOverhang</span> <span class="pre">100</span></code> is just fine). If you changed this number, remember this for the next steps as you will need to provide it again if changed here.</div>
<div class="line">- This may take awhile, and as we will discuss later, you may want to run these steps on a supercomputer, but this will serve as a preliminary guide for now.</div>
<div class="line">- One final consideration – if we are dealing with an organism with a smaller genome size, we will want to provide the <code class="xref py py-data docutils literal notranslate"><span class="pre">--genome_size</span></code> parameter with the the number of nucleotides in the organism’s genome. If you change this parameter in this step, you will need to provide the parameter and value in the <code class="xref py py-data docutils literal notranslate"><span class="pre">align</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">riboseq</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">seRNAseq</span></code>, and <code class="xref py py-data docutils literal notranslate"><span class="pre">seRNAseq</span></code> modules.</div>
</div>
</div>
<div class="section" id="process-raw-sequencing-files">
<h2>Process Raw Sequencing Files<a class="headerlink" href="#process-raw-sequencing-files" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">- Now let’s get our raw data::</div>
<div class="line">1. Make a new folder, something called <code class="xref py py-data docutils literal notranslate"><span class="pre">raw_data</span></code> or whatever you like and place your data there.</div>
<div class="line">2. Make sure the files follow proper naming conventions (see naming conventions at <a class="reference internal" href="general-usage.html#general-link"><span class="std std-ref">General Usage</span></a>)</div>
<div class="line">3. Now let’s process the data</div>
<div class="line">4. Let’s also create a folder called something like <code class="xref py py-data docutils literal notranslate"><span class="pre">output</span></code></div>
<div class="line">5. Also, make sure you have the 3’ adapter sequence handy used when generating your sequencing library</div>
<div class="line">6. We’ll feed the program the new GTF file that contains only longest transcript, protein coding, truncated references generating in the reference curation step</div>
<div class="line">7. We’ll give the experiment a name and also specify what <a class="reference external" href="https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/">method of sample normalization</a> we want performed on the count data</div>
<div class="line">8. We also need to specify the <code class="xref py py-data docutils literal notranslate"><span class="pre">--sjdbOverhang</span></code> amount we fed into the reference curation step, so in this case we will use <code class="xref py py-data docutils literal notranslate"><span class="pre">--sjdbOverhang</span> <span class="pre">49</span></code></div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe riboseq --input raw_data/ <span class="se">\</span>
                    --output output/ <span class="se">\</span>
                    --reference reference_folder/ <span class="se">\</span>
                    --gtf reference_folder/transcripts_LCT.gtf
                    --experiment riboseq_test
                    --adapter CTGTAGGCACCATCAAT
                    --method RPKM
                    --sjdbOverhang <span class="m">49</span>

<span class="c1">### or ###</span>

$ xpresspipe build

<span class="c1">### And then choose the appropriate pipeline to build</span>
</pre></div>
</div>
<div class="line-block">
<div class="line">- If you are running a lot of files, especially for human samples, this may take a lot of time. We recommend running this on some kind of server. A situation like yeast with few samples may be feasible to run on a personal computer, but will likely also take some time.</div>
</div>
<div class="section" id="sequencing-metrics">
<h3>Sequencing Metrics<a class="headerlink" href="#sequencing-metrics" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line">In your output folder, you will see a file named <code class="xref py py-data docutils literal notranslate"><span class="pre">riboseq_test_multiqc_report.html</span></code>. This file will compile the statistics from each processing step of the pipeline for each sample file you provided as input. Things like read quality, mapping, and quantification statistics can be found here. Just double-click the file or execute the following command to open in your default browser window.</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ open riboseq_test_multiqc_report.html
</pre></div>
</div>
</div>
<div class="section" id="library-complexity">
<h3>Library Complexity<a class="headerlink" href="#library-complexity" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line">Within the <code class="xref py py-data docutils literal notranslate"><span class="pre">complexity</span></code> directory in your output folder, you will find summary PDFs for all samples processed analyzing library complexity of each sample.</div>
</div>
</div>
<div class="section" id="metagene-analysis">
<h3>Metagene Analysis<a class="headerlink" href="#metagene-analysis" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line">Within the <code class="xref py py-data docutils literal notranslate"><span class="pre">metagene</span></code> directory in your output folder, you will find summary PDFs for all samples processed analyzing the metagene profile of each sample.</div>
</div>
</div>
<div class="section" id="periodicity-ribosome-profiling">
<h3>Periodicity (Ribosome Profiling)<a class="headerlink" href="#periodicity-ribosome-profiling" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line">Within the <code class="xref py py-data docutils literal notranslate"><span class="pre">periodicity</span></code> directory in your output folder, you will find summary PDFs for all samples processed analyzing ribosome periodicity of each of each sample containing reads 28-30nt.</div>
</div>
</div>
<div class="section" id="count-data-and-downstream-analysis">
<h3>Count Data and Downstream Analysis<a class="headerlink" href="#count-data-and-downstream-analysis" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line">Within the <code class="xref py py-data docutils literal notranslate"><span class="pre">counts</span></code> directory in your output folder, you will find individual counts tables for each sample, as well as compiled tables for each sample that was processed.</div>
</div>
</div>
</div>
<div class="section" id="supercomputing">
<h2>Supercomputing<a class="headerlink" href="#supercomputing" title="Permalink to this headline">¶</a></h2>
<div class="section" id="install">
<h3>Install<a class="headerlink" href="#install" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line">- Much of the same commands will be performed as above, aside from a couple exceptions:</div>
<div class="line">1. When installing XPRESSpipe, you need to provide a location for personal storage of the software:</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ python setup.py install --prefix ~/.local/bin
</pre></div>
</div>
<div class="line-block">
<div class="line">2. Add this path to your <code class="xref py py-data docutils literal notranslate"><span class="pre">$PATH</span></code>:</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ <span class="nb">echo</span> <span class="s1">&#39;export PATH=&quot;~/.local/bin:$PATH&quot;&#39;</span> &gt;&gt; ~/.bash_profile
$ <span class="nb">echo</span> <span class="s1">&#39;export PYTHONPATH=&quot;/uufs/chpc.utah.edu/common/home/u0690617/.local/bin/lib/python3.7/site-packages&quot;&#39;</span> &gt;&gt; ~/.bash_profile
</pre></div>
</div>
<div class="line-block">
<div class="line">3. Let’s test this to make sure everything is operating properly:</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ <span class="nb">cd</span> ~/
$ xpresspipe <span class="nb">test</span>
</pre></div>
</div>
</div>
<div class="section" id="run-data">
<h3>Run Data<a class="headerlink" href="#run-data" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line">- The commands here are the same as above, but likely the method of execution will be different. A lot of supercomputing clusters manage job submission through a system called <a class="reference external" href="https://www.youtube.com/watch?v=RpkAyFI05yY">SLURM</a>. Likely, the supercomputing cluster you are running your data on will have instructions for how to use this, but briefly, here is an example batch script (should end in the suffix <code class="xref py py-data docutils literal notranslate"><span class="pre">sh</span></code>):</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span><span class="ch">#!/bin/bash</span>
<span class="c1">#SBATCH --time=72:00:00</span>
<span class="c1">#SBATCH --nodes=1</span>
<span class="c1">#SBATCH -o /scratch/general/lustre/$USER/slurmjob-%j</span>
<span class="c1">#SBATCH --partition=this_cluster_has_no_name</span>

<span class="c1">#set up the temporary directory</span>
<span class="nv">SCRDIR</span><span class="o">=</span>/scratch/general/lustre/<span class="nv">$USER</span>/<span class="nv">$SLURM_JOBID</span>
mkdir -p <span class="nv">$SCRDIR</span>

<span class="c1"># Provide location of raw data and parent reference directory</span>
<span class="nv">SRA</span><span class="o">=</span>/scratch/general/lustre/<span class="nv">$USER</span>/files/your_favorite_experiment_goes_here
<span class="nv">REF</span><span class="o">=</span>/scratch/general/lustre/<span class="nv">$USER</span>/references/fantastic_creature_reference

<span class="c1"># Send raw data to your Scratch directory</span>
mkdir <span class="nv">$SCRDIR</span>/input
cp <span class="nv">$SRA</span>/*.fastq <span class="nv">$SCRDIR</span>/input/.

<span class="c1"># Make an output directory</span>
mkdir <span class="nv">$SCRDIR</span>/output
<span class="nb">cd</span> <span class="nv">$SCRDIR</span>/.

xpresspipe riboseq -i <span class="nv">$SCRDIR</span>/input -o <span class="nv">$SCRDIR</span>/output/ -r <span class="nv">$REF</span> --gtf <span class="nv">$REF</span>/transcripts_CT.gtf -e this_is_a_test -a CTGTAGGCACCATCAAT --sjdbOverhang
</pre></div>
</div>
<div class="line-block">
<div class="line">- To queue this script into the job pool, you would do the following:</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ sbatch my_batch_script.sh
</pre></div>
</div>
<div class="line-block">
<div class="line">- To monitor the progress of your job, execute the following:</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ watch -n1 squeue -u <span class="nv">$USER</span>
</pre></div>
</div>
<div class="line-block">
<div class="line">- After the job is finished, you can export the data as shown in the next section.</div>
</div>
</div>
<div class="section" id="explore-the-data">
<h3>Explore the Data<a class="headerlink" href="#explore-the-data" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line">Once the data is finished processing, we can start exploring the output. Explanations each quality control analysis can be found in the <a class="reference internal" href="analysis.html#analysis-link"><span class="std std-ref">Analysis</span></a> section of the documentation.</div>
<div class="line">In order to get the data from a HPC to your personal computer, you can use a command like the following:</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ <span class="nb">cd</span> ~/Desktop <span class="c1"># Or any other location where you want to store and analyze the data</span>
$ scp USERNAME@this_cluster_has_no_name.chpc.university.edu:/full/path/to/files/file_name.suffix ./
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="installation.html" class="btn btn-neutral float-right" title="Installation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="quickstart.html" class="btn btn-neutral float-left" title="Quickstart" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Jordan A. Berg

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>