

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Quality Control &mdash; XPRESSpipe 0.2.4-beta documentation</title>
  

  
  
    <link rel="shortcut icon" href="../_static/xpressyourself.png"/>
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Analysis" href="analysis.html" />
    <link rel="prev" title="Ribosome Profiling Pipeline" href="riboseq.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> XPRESSpipe
          

          
            
            <img src="../_static/xpressyourself.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">XPRESSpipe</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="beginner.html">Beginner’s Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="general-usage.html">General Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference-building.html">Curating References</a></li>
<li class="toctree-l1"><a class="reference internal" href="seRNAseq.html">Single-End RNA-seq Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="peRNAseq.html">Paired-End RNA-seq Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="riboseq.html">Ribosome Profiling Pipeline</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Quality Control</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#read-distribution-analysis">Read Distribution Analysis</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#arguments">Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#examples">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#metagene-analysis">Metagene Analysis</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id1">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#intron-collapsed-gene-coverage-analysis">Intron-collapsed Gene Coverage Analysis</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id2">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#codon-periodicity-analysis">Codon Periodicity Analysis</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id3">Examples</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="analysis.html">Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="trim.html">Read Pre-Processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="align.html">Alignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="count.html">Read Quantification</a></li>
<li class="toctree-l1"><a class="reference internal" href="normalization.html">Normalize</a></li>
<li class="toctree-l1"><a class="reference internal" href="other-features.html">Other Features</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="updates.html">Updates</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">XPRESSpipe</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Quality Control</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/content/quality-control.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="quality-control">
<h1>Quality Control<a class="headerlink" href="#quality-control" title="Permalink to this headline">¶</a></h1>
<div class="section" id="read-distribution-analysis">
<h2>Read Distribution Analysis<a class="headerlink" href="#read-distribution-analysis" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">When performing RNA-seq, your sequencing library population is important to assess to ensure a quality sequencing run. Unexpected populations can be indicative of RNA degradation or other effects. In ribosome profiling, the expected footprint size is ~28-30 nucleotides, so you would expect a peak in this region when running your analysis. The following module will run read distribution analysis for all <code class="xref py py-data docutils literal notranslate"><span class="pre">fastq</span></code> samples within a given directory. It is recommended this analysis be performed on trimmed reads to clean up adapters and get the true distribution of sequence reads in the library. When this is run within the pipeline, it will analyze just the post-trimming <code class="xref py py-data docutils literal notranslate"><span class="pre">fastq</span></code> files.</div>
</div>
<div class="line-block">
<div class="line">Additionally, if running one of XPRESSpipe’s pipelines, you can refer to the MultiQC <code class="xref py py-data docutils literal notranslate"><span class="pre">html</span></code> file for general summary statistics, which include read length distributions for all samples.</div>
</div>
<div class="section" id="arguments">
<h3>Arguments<a class="headerlink" href="#arguments" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line">The help menu can be accessed by calling the following from the command line:</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe readDistribution --help
</pre></div>
</div>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Required Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-i</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--input</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to input directory of trimmed fastq (or untrimmed fastq) files</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-o</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--output</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to output directory</p></td>
</tr>
</tbody>
</table>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Required Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-t</span> <span class="pre">&lt;SE</span> <span class="pre">or</span> <span class="pre">PE&gt;,</span> <span class="pre">--type</span> <span class="pre">&lt;SE</span> <span class="pre">or</span> <span class="pre">PE&gt;</span></code></p></td>
<td><p>Sequencing type (“SE” for single-end, “PE” for paired-end)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-e</span> <span class="pre">&lt;experiment_name&gt;,</span> <span class="pre">--experiment</span> <span class="pre">&lt;experiment_name&gt;</span></code></p></td>
<td><p>Experiment name</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-m</span></code></p></td>
<td><p>Number of max processors to use for tasks (default: No limit)</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="examples">
<h3>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line"><strong>Example 1 – Analyze read distributions from ribosome profiling libraries</strong></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>$ xpresspipe readDistribution -i riboprof_out/trimmed_fastq -o riboprof_out -e se_test
</pre></div>
</div>
<a class="reference internal image-reference" href="../_images/se_test_fastqc_summary.png"><img alt="../_images/se_test_fastqc_summary.png" src="../_images/se_test_fastqc_summary.png" style="width: 450px;" /></a>
</div>
</div>
<div class="section" id="metagene-analysis">
<h2>Metagene Analysis<a class="headerlink" href="#metagene-analysis" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">Analyze each sequencing sample to ensure equal distribution of reads across all transcripts. Can be useful in identifying 5’ or 3’ biases in sequence preparation.</div>
<div class="line">Requires a transcriptome-mapped BAM files, which can be output by <a class="reference external" href="https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf">STAR</a> and are automatically output during any XPRESSpipe alignment run.</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe metagene --help
</pre></div>
</div>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Required Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-i</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--input</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to input directory of transcriptome-mapped BAM files</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-o</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--output</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to output directory</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-g</span> <span class="pre">&lt;/path/transcripts.gtf&gt;</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--gtf</span> <span class="pre">&lt;/path/transcripts.gtf&gt;</span></code></p></td>
<td><p>Path and file name to un-modified reference GTF</p></td>
</tr>
</tbody>
</table>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Optional Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-e</span> <span class="pre">&lt;experiment_name&gt;,</span> <span class="pre">--experiment</span> <span class="pre">&lt;experiment_name&gt;</span></code></p></td>
<td><p>Experiment name</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--feature_type</span> <span class="pre">&lt;feature_type&gt;</span></code></p></td>
<td><p>Specify feature type (3rd column in GTF file) to be used in calculating metagene coverage (default: exon; alternative: CDS)</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--bam_suffix</span> <span class="pre">&lt;suffix&gt;</span></code></p></td>
<td><p>Change from default suffix of toTranscriptome.out.bam if transcriptome-mapped files were processed outside of XPRESSpipe</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-m</span> <span class="pre">&lt;processors&gt;,</span> <span class="pre">--max_processors</span> <span class="pre">&lt;processors&gt;</span></code></p></td>
<td><p>Number of max processors to use for tasks (default: No limit)</p></td>
</tr>
</tbody>
</table>
<div class="section" id="id1">
<h3>Examples<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line"><strong>Example 1 – Analyze metagene profiles of sequence libraries</strong></div>
<div class="line">- Use default transcript reference (maps to all transcripts, even if non-coding)</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>$ xpresspipe metagene -i riboprof_out/alignments/ -o riboprof_out -g se_reference/transcripts.gtf -e se_test
</pre></div>
</div>
<a class="reference internal image-reference" href="../_images/se_test_metagene_summary.png"><img alt="../_images/se_test_metagene_summary.png" src="../_images/se_test_metagene_summary.png" style="width: 450px;" /></a>
<p>NOTE: As you can appreciate, there are systematic 5’ biases in these library preparations. A good RNA-seq library should generally have even coverage across all transcripts.</p>
</div>
</div>
<div class="section" id="intron-collapsed-gene-coverage-analysis">
<h2>Intron-collapsed Gene Coverage Analysis<a class="headerlink" href="#intron-collapsed-gene-coverage-analysis" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">Plot the coverage of a given gene for a sample or set of samples with introns collapsed.</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe geneCoverage --help
</pre></div>
</div>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Required Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-i</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--input</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to input directory of transcriptome-aligned BAM files</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-o</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--output</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to output directory</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-g</span> <span class="pre">&lt;/path/transcripts.gtf&gt;</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--gtf</span> <span class="pre">&lt;/path/transcripts.gtf&gt;</span></code></p></td>
<td><p>Path and file name to reference GTF</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-n</span> <span class="pre">&lt;gene_name&gt;,</span> <span class="pre">--gene_name</span> <span class="pre">&lt;gene_name&gt;</span></code></p></td>
<td><p>Gene name (case sensitive)</p></td>
</tr>
</tbody>
</table>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Optional Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-e</span> <span class="pre">&lt;experiment_name&gt;,</span> <span class="pre">--experiment</span> <span class="pre">&lt;experiment_name&gt;</span></code></p></td>
<td><p>Experiment name to save output summaries as</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--bam_suffix</span> <span class="pre">&lt;suffix&gt;</span></code></p></td>
<td><p>Change from default suffix of toTranscriptome.out.bam if using a different BAM file</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--type</span> <span class="pre">&lt;type&gt;</span></code></p></td>
<td><p>Record type to map across (i.e. “exon”, “CDS”) (case-sensitive)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--samples</span> <span class="pre">&lt;sample_list&gt;</span> <span class="pre">[&lt;sample_list&gt;</span> <span class="pre">...]</span></code></p></td>
<td><p>Provide a space-separated list of sample names to include in analysis (will only include those listed, and will plot in the order listed)</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--sample_names</span> <span class="pre">&lt;suffix&gt;</span></code></p></td>
<td><p>Provide a space-separated list of sample names to use for labels</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--plot_color</span> <span class="pre">&lt;color&gt;</span></code></p></td>
<td><p>Indicate plotting color</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-m</span> <span class="pre">&lt;processors&gt;,</span> <span class="pre">--max_processors</span> <span class="pre">&lt;processors&gt;</span></code></p></td>
<td><p>Number of max processors to use for tasks (default: No limit)</p></td>
</tr>
</tbody>
</table>
<div class="section" id="id2">
<h3>Examples<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line"><strong>Example 1 – Analyze gene coverage profile of sequence libraries</strong></div>
<div class="line">- Use default transcript reference (will generate a longest transcript-only reference)</div>
<div class="line">- Analyze SLC1A1</div>
<div class="line">- Analyze along chosen record type (default: exon, but could also use CDS if looking at ribosome profiling data)</div>
<div class="line">- Analyzing BAM files ending in <code class="xref py py-data docutils literal notranslate"><span class="pre">sort.bam</span></code></div>
<div class="line">- Specifying names to use in plotting – if not using <code class="xref py py-data docutils literal notranslate"><span class="pre">--samples</span></code>, these files will be plotted alphabetically, so the listed order should also be alphabetical. If using <code class="xref py py-data docutils literal notranslate"><span class="pre">--samples</span></code>, need to specify names in the same order you provided for this argument.</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>$ xpresspipe geneCoverage -i /path/to/bam_files -o ./ -g /path/to/reference.gtf \
  -n SLC1A1 --type exon --bam_suffix .sort.bam \
  --sample_names SRR1795425 SRR1795433 SRR1795435 SRR1795437
</pre></div>
</div>
<a class="reference internal image-reference" href="../_images/geneCoverage_IGV_comparison.png"><img alt="../_images/geneCoverage_IGV_comparison.png" src="../_images/geneCoverage_IGV_comparison.png" style="width: 750px;" /></a>
<p>NOTE: The coverage estimations use a 20 nt rolling window mean method to smoothen the coverage plots. In both A and B in the image above, the top plot was generated with IGV (<a class="reference external" href="https://software.broadinstitute.org/software/igv/">https://software.broadinstitute.org/software/igv/</a>) and the bottom with <code class="xref py py-data docutils literal notranslate"><span class="pre">xpresspipe</span> <span class="pre">geneCoverage</span></code>. Green boxes show approximately the same region for comparison.</p>
</div>
</div>
<div class="section" id="codon-periodicity-analysis">
<h2>Codon Periodicity Analysis<a class="headerlink" href="#codon-periodicity-analysis" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">Analyze periodicity of most abundant read length. Useful in ribosome profiling samples for identifying that ribosomes are taking the expected 3 nucleotide steps along a transcript. If this is not apparent from the analysis, it may be indicative of poor sequence coverage of the ribosome profiling libraries.</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe periodicity --help
</pre></div>
</div>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Required Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-i</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--input</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to input directory of transcriptome-aligned BAM files</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-o</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--output</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to output directory</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-g</span> <span class="pre">&lt;/path/transcripts.gtf&gt;</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--gtf</span> <span class="pre">&lt;/path/transcripts.gtf&gt;</span></code></p></td>
<td><p>Path and file name to reference GTF</p></td>
</tr>
</tbody>
</table>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Optional Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-e</span> <span class="pre">&lt;experiment_name&gt;,</span> <span class="pre">--experiment</span> <span class="pre">&lt;experiment_name&gt;</span></code></p></td>
<td><p>Experiment name to save output summaries as</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--bam_suffix</span> <span class="pre">&lt;suffix&gt;</span></code></p></td>
<td><p>Change from default suffix of toTranscriptome.out.bam if using a different BAM file</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-m</span> <span class="pre">&lt;processors&gt;,</span> <span class="pre">--max_processors</span> <span class="pre">&lt;processors&gt;</span></code></p></td>
<td><p>Number of max processors to use for tasks (default: No limit)</p></td>
</tr>
</tbody>
</table>
<div class="section" id="id3">
<h3>Examples<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line"><strong>Example 1 – Analyze periodicity from ribosome profiling libraries</strong></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>$ xpresspipe periodicity -i riboprof_out/alignments/ -o riboprof_out -g se_reference/transcripts.gtf -e se_test
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="analysis.html" class="btn btn-neutral float-right" title="Analysis" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="riboseq.html" class="btn btn-neutral float-left" title="Ribosome Profiling Pipeline" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Jordan A. Berg

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>