

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Curating References &mdash; XPRESSpipe 0.2.4-beta documentation</title>
  

  
  
    <link rel="shortcut icon" href="../_static/xpressyourself.png"/>
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Single-End RNA-seq Pipeline" href="seRNAseq.html" />
    <link rel="prev" title="General Usage" href="general-usage.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> XPRESSpipe
          

          
            
            <img src="../_static/xpressyourself.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">XPRESSpipe</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="beginner.html">Beginner’s Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="general-usage.html">General Usage</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Curating References</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#xpresspipe-reference-requirements">XPRESSpipe Reference Requirements</a></li>
<li class="toctree-l2"><a class="reference internal" href="#get-sequence-files">Get Sequence Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#perform-full-reference-curation">Perform Full Reference Curation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#arguments">Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#examples">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#star-reference-curation">STAR Reference Curation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id1">Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id2">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#transcript-reference-modification">Transcript Reference Modification</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id3">Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id4">Examples</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="seRNAseq.html">Single-End RNA-seq Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="peRNAseq.html">Paired-End RNA-seq Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="riboseq.html">Ribosome Profiling Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="quality-control.html">Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis.html">Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="trim.html">Read Pre-Processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="align.html">Alignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="count.html">Read Quantification</a></li>
<li class="toctree-l1"><a class="reference internal" href="normalization.html">Normalize</a></li>
<li class="toctree-l1"><a class="reference internal" href="other-features.html">Other Features</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="updates.html">Updates</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">XPRESSpipe</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Curating References</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/content/reference-building.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="curating-references">
<h1>Curating References<a class="headerlink" href="#curating-references" title="Permalink to this headline">¶</a></h1>
<div class="line-block">
<div class="line">In order to quantify transcription levels from RNA-Seq data, reads must be mapped to a reference genome or transcriptome. While there are multiple alignment software packages available, XPRESSpipe performs this step using a current version of <a class="reference external" href="https://github.com/alexdobin/STAR">STAR</a> for several reasons:</div>
<div class="line">- <strong>Splice Junction Aware</strong>: STAR is capable of mapping reads spanning a splice junction, where more traditional packages, such as Bowtie, are incapable of doing so and are better suited for tasks such as genome alignment.</div>
<div class="line">- <strong>Performance</strong>: While computationally greedy (a human genome alignment requires upwards of 30 Gb RAM), the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5792058/">performance and accuracy is excellent compared to the majority of other splice-aware aligners currently available</a></div>
<div class="line">- <strong>Standard</strong>: The foundation of the pipeline used in XPRESSpipe is based in the <a class="reference external" href="https://docs.gdc.cancer.gov/Data/Bioinformatics_Pipelines/Expression_mRNA_Pipeline/">TCGA</a> standards for RNA-Seq alignment. This method utilizes a guided or 2-pass alignment program. In the guided alignment, a GTF with annotated splice junctions is used to guide the alignments over splice juntions. In the 2-pass alignment, reads are mapped across the genome to identify novel splice junctions. These new annotations are then incorporated into the reference index and reads are re-aligned with this new reference. While more time-intensive, this step can aid in aligning across these junctions, especially in organisms where the transcriptome is not as well annotated. If mapping to a well-documented organism, this step can be forgone and STAR will use the GTF annotations to determine intronic regions of transcripts for read mapping.</div>
</div>
<div class="section" id="xpresspipe-reference-requirements">
<h2>XPRESSpipe Reference Requirements<a class="headerlink" href="#xpresspipe-reference-requirements" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">An XPRESSpipe compatible reference directory must meet some requirements:</div>
<div class="line">- All chromosomal genome fasta files are in their own directory within the parent reference directory. If a FASTA file with all chromosomes combined is available for your organism, this can be provided, but must be in its own directory.</div>
<div class="line">- A sub-directory, named <code class="xref py py-data docutils literal notranslate"><span class="pre">genome</span></code>, contains the STAR reference files. If <code class="xref py py-data docutils literal notranslate"><span class="pre">createReference</span></code> is used to curate the reference, and the parent reference directory was provided as output location, this directory creation and file formatting will be handled automatically by XPRESSpipe.</div>
<div class="line">- A transcript reference (GTF), is located in the reference parent directory and is named <code class="xref py py-data docutils literal notranslate"><span class="pre">transcripts.gtf</span></code>. If a coding-only or truncated reference GTFs are desired for read quantification, these should also be in this directory (<code class="xref py py-data docutils literal notranslate"><span class="pre">truncate</span></code> will handle file naming and formatting so long as the output location is specified as this parent directory). This file will then need to be specified within an XPRESSpipe pipeline.</div>
<div class="line"><strong>A completed reference directory can be created that follows these requirements by creating a directory, placing the transcripts.gtf and genomic chromosome fasta files in the parent directory and running :data:`curateReference` as described below</strong></div>
</div>
</div>
<div class="section" id="get-sequence-files">
<h2>Get Sequence Files<a class="headerlink" href="#get-sequence-files" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">The following is an example of how to get the reference files needed for generating a human reference:</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ mkdir human_reference
$ mkdir human_reference/genome_fasta
$ <span class="nb">cd</span> human_reference/
$ curl ftp://ftp.ensembl.org/pub/release-95/gtf/homo_sapiens/Homo_sapiens.GRCh38.95.gtf.gz -o transcripts.gtf.gz
$ <span class="k">for</span> i in <span class="m">1</span> <span class="m">2</span> <span class="m">3</span> <span class="m">4</span> <span class="m">5</span> <span class="m">6</span> <span class="m">7</span> <span class="m">8</span> <span class="m">9</span> <span class="m">10</span> <span class="m">11</span> <span class="m">12</span> <span class="m">13</span> <span class="m">14</span> <span class="m">15</span> <span class="m">16</span> <span class="m">17</span> <span class="m">18</span> <span class="m">19</span> <span class="m">20</span> <span class="m">21</span> <span class="m">22</span> X Y<span class="p">;</span> <span class="k">do</span> curl -O ftp://ftp.ensembl.org/pub/release-95/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.chromosome.<span class="si">${</span><span class="nv">i</span><span class="si">}</span>.fa.gz<span class="p">;</span> <span class="k">done</span>
$ gzip -d *.gz
$ mv *fasta genome_fasta
</pre></div>
</div>
<div class="line-block">
<div class="line">- The chromosome IDs may vary depending on your organism.</div>
</div>
</div>
<div class="section" id="perform-full-reference-curation">
<h2>Perform Full Reference Curation<a class="headerlink" href="#perform-full-reference-curation" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">The following will create a XPRESSpipe-formatted reference directory containing all STAR reference files and transcript references needs for quantification and meta-analysis.</div>
<div class="line">A parent reference directory containing the transcripts.gtf file and all chromosomal genome fasta files must be present.</div>
<div class="line"><em>More details as to what each specific parameter is doing can be found in the sections below.</em></div>
</div>
<div class="section" id="arguments">
<h3>Arguments<a class="headerlink" href="#arguments" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line">The help menu can be accessed by calling the following from the command line:</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe curateReference --help
</pre></div>
</div>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Required Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-o</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--output</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to output directory</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-f</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--fasta</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to genome fasta files (file names should end in .fa, .fasta, or .txt and no other files should exist in the directory with similar extensions)</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-g</span> <span class="pre">&lt;/path/transcripts.gtf&gt;</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--gtf</span> <span class="pre">&lt;/path/transcripts.gtf&gt;</span></code></p></td>
<td><p>Path and file name to transcript reference file names ‘transcripts.gtf’</p></td>
</tr>
</tbody>
</table>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Optional Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-l,</span> <span class="pre">--longest_transcript</span></code></p></td>
<td><p>Provide argument to keep only longest transcript per gene record (RECOMMENDED)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-p,</span> <span class="pre">--protein_coding</span></code></p></td>
<td><p>Provide argument to keep only gene records annotated as protein coding genes</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-t,</span> <span class="pre">--truncate</span></code></p></td>
<td><p>Provide argument to truncate gene records</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--truncate_5prime</span></code></p></td>
<td><p>Amount to truncate from 5’ end of each transcript, requires –truncate argument (default: 45)</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--truncate_3prime</span></code></p></td>
<td><p>Amount to truncate from 3’ end of each transcript, requires –truncate argument (default: 15)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--sjdbOverhang</span> <span class="pre">&lt;value&gt;</span></code></p></td>
<td><p>Specify length of genomic sequences for constructing splice-aware reference. Ideal length is <code class="xref py py-data docutils literal notranslate"><span class="pre">read</span> <span class="pre">length</span> <span class="pre">-</span> <span class="pre">1</span></code>, so for 2x100bp paired-end reads, you would use 100 - 1 = 99. However, the default value of <code class="xref py py-data docutils literal notranslate"><span class="pre">100</span></code> should work in most cases</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--genome_size</span> <span class="pre">&lt;int&gt;</span></code></p></td>
<td><p>If mapping to an organism with a small genome, provide the length in nucleotides. If you are not sure your organism has a small genome, provide the number of bases and XPRESSpipe will decide if this parameter needs to be changed during runtime</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-m</span></code></p></td>
<td><p>Number of max processors to use for tasks (default: No limit)</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="examples">
<h3>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line"><strong>Example 1 – Create XPRESSpipe-formatted reference for single-end alignment:</strong></div>
<div class="line">- Creates a star reference for single-end read mapping (1x50bp reads)</div>
<div class="line">- Keeps the longest transcript for each gene record</div>
<div class="line">- Keeps only protein_coding annotated transcripts</div>
<div class="line">- Truncates the first 45 nucleotides from the first exon of every transcript (default)</div>
<div class="line">- Truncates the last 15 nucleotides from the last exon of every transcript (default)</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe curateReference -o /path/to/se/ref/ -f /path/to/se/ref/ -g /path/to/se/ref/transcripts.gtf --longest_transcript --protein_coding --truncate --sjdbOverhang <span class="m">49</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><strong>Example 2 – Create refFlat files:</strong></div>
<div class="line">- Creates a star reference for paired-end read mapping (2x100bp reads)</div>
<div class="line">- No modifications are made to the GTF file</div>
<div class="line">- Processes are limited to 10 cores</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe curateReference -o /path/to/pe/ref/ -f /path/to/pe/ref/ -g /path/to/pe/ref/transcripts.gtf -m <span class="m">10</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="star-reference-curation">
<h2>STAR Reference Curation<a class="headerlink" href="#star-reference-curation" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">The following creates a STAR reference compatible with XPRESSpipe. These files are output in a directory created during curation called <code class="xref py py-data docutils literal notranslate"><span class="pre">genome</span></code> in the specified <code class="xref py py-data docutils literal notranslate"><span class="pre">--output</span></code> directory.</div>
</div>
<div class="section" id="id1">
<h3>Arguments<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line">The help menu can be accessed by calling the following from the command line:</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe makeReference --help
</pre></div>
</div>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Required Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-o</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--output</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to output directory</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-f</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--fasta</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to genome fasta files (file names should end in .fa, .fasta, or .txt and no other files should exist in the directory with similar extensions)</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-g</span> <span class="pre">&lt;/path/transcripts.gtf&gt;</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--gtf</span> <span class="pre">&lt;/path/transcripts.gtf&gt;</span></code></p></td>
<td><p>Path and file name to transcript reference file names ‘transcripts.gtf (DO NOT USE MODIFIED GTF HERE)’</p></td>
</tr>
</tbody>
</table>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Optional Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--sjdbOverhang</span> <span class="pre">&lt;int&gt;</span></code></p></td>
<td><p>Specify length of genomic sequences for constructing splice-aware reference. Ideal length is <code class="xref py py-data docutils literal notranslate"><span class="pre">read</span> <span class="pre">length</span> <span class="pre">-</span> <span class="pre">1</span></code>, so for 2x100bp paired-end reads, you would use 100 - 1 = 99. However, the default value of <code class="xref py py-data docutils literal notranslate"><span class="pre">100</span></code> should work in most cases</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--genome_size</span> <span class="pre">&lt;int&gt;</span></code></p></td>
<td><p>If mapping to an organism with a small genome, provide the length in nucleotides. If you are not sure your organism has a small genome, provide the number of bases and XPRESSpipe will decide if this parameter needs to be changed during runtime</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-m</span></code></p></td>
<td><p>Number of max processors to use for tasks (default: No limit)</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id2">
<h3>Examples<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line"><strong>Example 1 – Create a single-end sequencing reference:</strong></div>
<div class="line">- Paths to output and location of genome fasta files for each chromosome are provided, as well as path and file name to transcripts.gtf file</div>
<div class="line">- Default number of threads are used for preparing reference</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe makeReference -o /path/to/reference/ -f /path/to/reference/ -g /path/to/reference/transcripts.gtf --sjdbOverhang <span class="m">49</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><strong>Example 2 – Create a paired-end sequencing reference:</strong></div>
<div class="line">- 12 threads are specified for reference creation</div>
<div class="line">- The as 2x100bp paired-end sequencing was used, the default value for <code class="xref py py-data docutils literal notranslate"><span class="pre">--sjdbOverhang</span></code> of <code class="xref py py-data docutils literal notranslate"><span class="pre">100</span></code> is appropriate in this case</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe makeReference -o /path/to/reference/ -f /path/to/reference/ -g /path/to/reference/transcripts.gtf -t <span class="m">12</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><strong>Example 3 – Create a single-end sequencing reference for Saccharomyces cerevisiae:</strong></div>
<div class="line">- Paths to output and location of genome fasta files for each chromosome are provided, as well as path and file name to transcripts.gtf file</div>
<div class="line">- Default number of threads are used for preparing reference</div>
<div class="line">- Genome size is specified</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe makeReference -o /path/to/reference/ -f /path/to/reference/ -g /path/to/reference/transcripts.gtf --sjdbOverhang <span class="m">49</span> --genome_size <span class="m">3000000</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="transcript-reference-modification">
<h2>Transcript Reference Modification<a class="headerlink" href="#transcript-reference-modification" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">At times, quantification of transcripts to a modified transcript reference is desirable. Below are some examples:</div>
<div class="line">1. As ribosomal RNA (rRNA) contamination is common in RNA-seq, even when a depletion step was performed prior to library preparation, it is sometimes desirable to not count these and other non-coding RNAs in the quantification and analysis.</div>
<div class="line">2. During ribosome profiling library preparation, where a 5’ and 3’ pile-up of ribosome footprints due to slow initation and termination kinetics of footprints is common, it is suggested to <a class="reference external" href="https://www.cell.com/cms/10.1016/j.celrep.2016.01.043/attachment/257faf34-ff8f-4071-a642-bfdb531c75b8/mmc1">exclude the first 45-50 nucleotides from the 5’ end and 15 nucleotides from the 3’ end of each transcript during quantification</a>. This command will automatically curate an Ensembl GTF to meet these demands for read quantification.</div>
<div class="line">3. Several genes encode multiple isoforms or transcripts. During quantification, many software packages for counting reads to genes consider a read mapping to multiple transcripts of the same gene as a multi-mapper. Unless interested in alternate isoform usage, it is recommended that transcriptome reference files only contain the longest transcript for each gene.</div>
<div class="line">The <code class="xref py py-data docutils literal notranslate"><span class="pre">modifyGTF</span></code> sub-module provides the ability to make the above-mentioned modifications to a GTF transcriptome reference file. The modified GTF file is output at the end and the filename is labeled with the modifications made. Truncations to each transcript reference are stranded-aware.</div>
</div>
<div class="section" id="id3">
<h3>Arguments<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line">The help menu can be accessed by calling the following from the command line:</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe modifyGTF --help
</pre></div>
</div>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Required Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-g</span> <span class="pre">&lt;/path/transcripts.gtf&gt;</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--gtf</span> <span class="pre">&lt;/path/transcripts.gtf&gt;</span></code></p></td>
<td><p>Path and file name to reference GTF</p></td>
</tr>
</tbody>
</table>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Optional Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-l,</span> <span class="pre">--longest_transcript</span></code></p></td>
<td><p>Provide argument to keep only longest transcript per gene record (RECOMMENDED)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-p,</span> <span class="pre">--protein_coding</span></code></p></td>
<td><p>Provide argument to keep only gene records annotated as protein coding genes</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-t,</span> <span class="pre">--truncate</span></code></p></td>
<td><p>Provide argument to truncate gene records</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--truncate_5prime</span></code></p></td>
<td><p>Amount to truncate from 5’ end of each transcript, requires –truncate argument (default: 45)</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--truncate_3prime</span></code></p></td>
<td><p>Amount to truncate from 3’ end of each transcript, requires –truncate argument (default: 15)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-m</span></code></p></td>
<td><p>Number of max processors to use for tasks (default: No limit)</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id4">
<h3>Examples<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line"><strong>Example 1 – Create longest transcript-only, protein coding-only, truncated reference:</strong></div>
<div class="line">- Keeps the longest transcript for each gene record</div>
<div class="line">- Keeps only protein_coding annotated transcripts</div>
<div class="line">- Truncates the first 45 nucleotides from the first exon of every transcript (default)</div>
<div class="line">- Truncates the last 15 nucleotides from the last exon of every transcript (default)</div>
<div class="line">- Each modification desired must be implicitly passed to the sub-module</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe modifyGTF -g /path/to/reference/transcripts.gtf --longest_transcript --protein_coding --truncate
</pre></div>
</div>
<span class="target" id="curate-link"></span></div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="seRNAseq.html" class="btn btn-neutral float-right" title="Single-End RNA-seq Pipeline" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="general-usage.html" class="btn btn-neutral float-left" title="General Usage" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Jordan A. Berg

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>