

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Ribosome Profiling Pipeline &mdash; XPRESSpipe 0.2.4-beta documentation</title>
  

  
  
    <link rel="shortcut icon" href="../_static/xpressyourself.png"/>
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Quality Control" href="quality-control.html" />
    <link rel="prev" title="Paired-End RNA-seq Pipeline" href="peRNAseq.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> XPRESSpipe
          

          
            
            <img src="../_static/xpressyourself.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">XPRESSpipe</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="beginner.html">Beginner’s Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="general-usage.html">General Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference-building.html">Curating References</a></li>
<li class="toctree-l1"><a class="reference internal" href="seRNAseq.html">Single-End RNA-seq Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="peRNAseq.html">Paired-End RNA-seq Pipeline</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Ribosome Profiling Pipeline</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#examples">Examples</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="quality-control.html">Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis.html">Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="trim.html">Read Pre-Processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="align.html">Alignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="count.html">Read Quantification</a></li>
<li class="toctree-l1"><a class="reference internal" href="normalization.html">Normalize</a></li>
<li class="toctree-l1"><a class="reference internal" href="other-features.html">Other Features</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="updates.html">Updates</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">XPRESSpipe</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Ribosome Profiling Pipeline</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/content/riboseq.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="ribosome-profiling-pipeline">
<h1>Ribosome Profiling Pipeline<a class="headerlink" href="#ribosome-profiling-pipeline" title="Permalink to this headline">¶</a></h1>
<div class="line-block">
<div class="line">The following pipeline will pre-process, align, and quality check ribosome profiling samples using the sub-modules discussed in other sections of this documentation. For more detailed information concerning these steps, please refer to the appropriate chapter for the step you are interested in.</div>
</div>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Required Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-i</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--input</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to input directory – if paired-end, file names should be exactly the same except for <code class="xref py py-data docutils literal notranslate"><span class="pre">r1/r2.fastq</span></code> or similar suffix</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-o</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--output</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to output directory</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-r</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--reference</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to parent organism reference directory</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-g</span> <span class="pre">&lt;/path/transcripts.gtf&gt;</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--gtf</span> <span class="pre">&lt;/path/transcripts.gtf&gt;</span></code></p></td>
<td><p>Path and file name to GTF used for alignment quantification (only used for HTSeq quantification)</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-e</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--experiment</span></code></p></td>
<td><p>Experiment name</p></td>
</tr>
</tbody>
</table>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Optional Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--two-pass</span></code></p></td>
<td><p>Use a two-pass STAR alignment for novel splice junction discovery</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-a</span> <span class="pre">&lt;adapter1</span> <span class="pre">...&gt;</span> <span class="pre">[&lt;adapter1</span> <span class="pre">...&gt;</span> <span class="pre">...]</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--adapter</span> <span class="pre">&lt;adapter1</span> <span class="pre">...&gt;</span> <span class="pre">[&lt;adapter1</span> <span class="pre">...&gt;</span> <span class="pre">...]</span></code></p></td>
<td><p>Specify adapter(s) in list of strings – for single-end, only provide one adapter – if <code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code> are provided, software will attempt to auto-detect adapters – if “POLYX” is provided as a single string in the list, polyX adapters will be trimmed. If you want to auto-detect adapters in for paired-end reads, provide <code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code> twice</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-q</span> <span class="pre">&lt;PHRED_value&gt;,</span> <span class="pre">--quality</span> <span class="pre">&lt;PHRED_value&gt;</span></code></p></td>
<td><p>PHRED read quality threshold (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">28</span></code>)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--min_length</span> <span class="pre">&lt;length_value&gt;</span></code></p></td>
<td><p>Minimum read length threshold to keep for reads (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">17</span></code>)</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--umi_location</span> <span class="pre">&lt;location&gt;</span></code></p></td>
<td><p>Provide parameter to process UMIs – provide location (see fastp documentation for more details, generally for single-end sequencing, you would provide ‘read1’ here; does not work with -a polyX option)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--umi_length</span> <span class="pre">&lt;length&gt;</span></code></p></td>
<td><p>Provide parameter to process UMIs – provide UMI length (must provide the –umi_location argument); does not work with -a polyX option)</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--no_multimappers&gt;</span></code></p></td>
<td><p>Include flag to remove multimapping reads to be output and used in downstream analyses</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--deduplicate</span></code></p></td>
<td><p>Include flag to quantify reads with de-duplication (will search for files with suffix <code class="xref py py-data docutils literal notranslate"><span class="pre">_dedupRemoved.bam</span></code>)</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--output_bed</span></code></p></td>
<td><p>Include flag to output BED files for each aligned file</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-c</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--quantification_method</span></code></p></td>
<td><p>Specify quantification method (default: htseq; other option: cufflinks. If using Cufflinks, no downstream sample normalization is required)</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--feature_type</span> <span class="pre">&lt;feature&gt;</span></code></p></td>
<td><p>Specify feature type (3rd column in GTF file) to be used if quantifying with htseq (default: CDS)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--stranded</span> <span class="pre">&lt;fr-unstranded/fr-firststrand</span></code> <code class="xref py py-data docutils literal notranslate"><span class="pre">/fr-secondstrand||no/yes&gt;</span></code></p></td>
<td><p>Specify whether library preparation was stranded (Options before || correspond with Cufflinks inputs, options after correspond with htseq inputs)</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--method</span> <span class="pre">&lt;RPM,</span> <span class="pre">RPKM,</span> <span class="pre">FPKM,</span> <span class="pre">TPM&gt;</span></code></p></td>
<td><p>Normalization method to perform (options: “RPM”, “TPM”, “RPKM”, “FPKM”) – if using either TPM, RPKM, or FPKM, a GTF reference file must be included</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--vcf</span> <span class="pre">&lt;/path/to/file.vcf&gt;</span></code></p></td>
<td><p>Provide full path and file name to VCF file if you would like detect personal variants overlapping alignments</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--batch</span> <span class="pre">&lt;/path/filename.tsv&gt;</span></code></p></td>
<td><p>Include path and filename of dataframe with batch normalization parameters</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--sjdbOverhang</span> <span class="pre">&lt;sjdbOverhang_amount&gt;</span></code></p></td>
<td><p>Specify length of genomic sequences for constructing splice-aware reference. Ideal length is <code class="xref py py-data docutils literal notranslate"><span class="pre">read</span> <span class="pre">length</span> <span class="pre">-</span> <span class="pre">1</span></code>, so for 2x100bp paired-end reads, you would use 100 - 1 = 99. However, the default value of <code class="xref py py-data docutils literal notranslate"><span class="pre">100</span></code> should work in most cases</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--mismatchRatio</span> <span class="pre">&lt;mismatchRatio&gt;</span></code></p></td>
<td><p>Alignment ratio of mismatches to mapped length is less than this value. See STAR documentation for more information on setting this parameter</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--seedSearchStartLmax</span> <span class="pre">&lt;seedSearchStartLmax&gt;</span></code></p></td>
<td><p>Adjusting this parameter by providing a lower number will improve mapping sensitivity (recommended value = 15 for reads ~ 25 nts). See STAR documentation for more information on setting this parameter</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">genome_size</span></code></p></td>
<td><p>Only needs to be changed if this argument was provided curing reference building AND using a two-pass alignment. This should be the length of the organism’s genome in nucleotides</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-m</span></code></p></td>
<td><p>Number of max processors to use for tasks (default: No limit)</p></td>
</tr>
</tbody>
</table>
<div class="line-block">
<div class="line">Run the following for more details:</div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>$ xpresspipe riboseq --help
</pre></div>
</div>
<div class="section" id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line"><strong>Example 1 – Run pipeline on ribosome profiling sample files</strong></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>$ xpresspipe riboseq \
              -i riboprof_test \
              -o ribopipe_out \
              -r se_reference \
              --gtf se_reference/transcript_CT.gtf \
              -e riboprof_test \
              -a CTGTAGGCACCATCAAT \
              --method RPM \
              --sjdbOverhang 49
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="quality-control.html" class="btn btn-neutral float-right" title="Quality Control" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="peRNAseq.html" class="btn btn-neutral float-left" title="Paired-End RNA-seq Pipeline" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Jordan A. Berg

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>