

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Read Pre-Processing &mdash; XPRESSpipe 0.2.4-beta documentation</title>
  

  
  
    <link rel="shortcut icon" href="../_static/xpressyourself.png"/>
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Alignment" href="align.html" />
    <link rel="prev" title="Analysis" href="analysis.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> XPRESSpipe
          

          
            
            <img src="../_static/xpressyourself.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">XPRESSpipe</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="beginner.html">Beginner’s Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="general-usage.html">General Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference-building.html">Curating References</a></li>
<li class="toctree-l1"><a class="reference internal" href="seRNAseq.html">Single-End RNA-seq Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="peRNAseq.html">Paired-End RNA-seq Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="riboseq.html">Ribosome Profiling Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="quality-control.html">Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis.html">Analysis</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Read Pre-Processing</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#read-trimming">Read Trimming</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#arguments">Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#examples">Examples</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="align.html">Alignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="count.html">Read Quantification</a></li>
<li class="toctree-l1"><a class="reference internal" href="normalization.html">Normalize</a></li>
<li class="toctree-l1"><a class="reference internal" href="other-features.html">Other Features</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="updates.html">Updates</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">XPRESSpipe</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Read Pre-Processing</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/content/trim.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="read-pre-processing">
<h1>Read Pre-Processing<a class="headerlink" href="#read-pre-processing" title="Permalink to this headline">¶</a></h1>
<div class="section" id="read-trimming">
<h2>Read Trimming<a class="headerlink" href="#read-trimming" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">Trimming is a necessary part of RNAseq data processing due to the technological limitations described below:</div>
<div class="line">- Inherent in RNA-seq library creation, RNA is fragmented and adapter sequences are ligated to the sequence. These adapters include information such as sample batch and act as a primer for the sequencer to recognize the fragment as something to analyze. However, these adapters, once sequenced, prevent alignment to a reference as large chunks of the fragment are synthetic sequence not found in the actual organism’s genome/transcriptome.</div>
<div class="line">- A sequencer’s job is to read a fragment base by base and determine the nucleotide species each step of the way. While the technology has greatly improved over the years, a probability of error remains. Mis-called bases can prevent proper alignment of the sequenced fragment to the reference. Therefore, it is important for low confidence base calls to be trimmed from each read.</div>
<div class="line"><br /></div>
<div class="line">Trimming is performed by <a class="reference external" href="https://github.com/OpenGene/fastp">fastp</a></div>
</div>
<div class="section" id="arguments">
<h3>Arguments<a class="headerlink" href="#arguments" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line">The help menu can be accessed by calling the following from the command line:</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe trim --help
</pre></div>
</div>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Required Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-i</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--input</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to input directory – if paired-end, file names should be exactly the same except for <code class="xref py py-data docutils literal notranslate"><span class="pre">r1/r2.fastq</span></code> or similar suffix</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-o</span> <span class="pre">&lt;path&gt;,</span> <span class="pre">--output</span> <span class="pre">&lt;path&gt;</span></code></p></td>
<td><p>Path to output directory</p></td>
</tr>
</tbody>
</table>
<table class="colwidths-given docutils align-center">
<colgroup>
<col style="width: 41%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Optional Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-a</span> <span class="pre">&lt;adapter1</span> <span class="pre">...&gt;</span> <span class="pre">[&lt;adapter1</span> <span class="pre">...&gt;</span> <span class="pre">...]</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">--adapter</span> <span class="pre">&lt;adapter1</span> <span class="pre">...&gt;</span> <span class="pre">[&lt;adapter1</span> <span class="pre">...&gt;</span> <span class="pre">...]</span></code></p></td>
<td><p>Specify adapter(s) in list of strings – for single-end, only provide one adapter – if <code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code> are provided, software will attempt to auto-detect adapters – if “POLYX” is provided as a single string in the list, polyX adapters will be trimmed. If you want to auto-detect adapters in for paired-end reads, provide <code class="xref py py-data docutils literal notranslate"><span class="pre">None</span></code> twice</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-q</span> <span class="pre">&lt;PHRED_value&gt;,</span> <span class="pre">--quality</span> <span class="pre">&lt;PHRED_value&gt;</span></code></p></td>
<td><p>PHRED read quality threshold (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">28</span></code>)</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--min_length</span> <span class="pre">&lt;length_value&gt;</span></code></p></td>
<td><p>Minimum read length threshold to keep for reads (default: <code class="xref py py-data docutils literal notranslate"><span class="pre">17</span></code>)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--umi_location</span> <span class="pre">&lt;location&gt;</span></code></p></td>
<td><p>Provide parameter to process UMIs – provide location (see fastp documentation for more details, generally for single-end sequencing, you would provide ‘read1’ here; does not work with -a polyX option)</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">--umi_length</span> <span class="pre">&lt;length&gt;</span></code></p></td>
<td><p>Provide parameter to process UMIs – provide UMI length (must provide the –umi_location argument); does not work with -a polyX option)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-data docutils literal notranslate"><span class="pre">-m</span></code></p></td>
<td><p>Number of max processors to use for tasks (default: Max)</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="examples">
<h3>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line"><strong>Example 1 – Trim ribosome profiling (or single-end) sequence data using default preferences:</strong></div>
<div class="line">- Raw reads are <code class="xref py py-data docutils literal notranslate"><span class="pre">fastq</span></code>-like and found in the <code class="xref py py-data docutils literal notranslate"><span class="pre">-i</span> <span class="pre">riboprof_test/</span></code> directory. Can be uncompressed or compressed via <code class="xref py py-data docutils literal notranslate"><span class="pre">gz</span></code> or <code class="xref py py-data docutils literal notranslate"><span class="pre">zip</span></code></div>
<div class="line">- A general output directory has been created, <code class="xref py py-data docutils literal notranslate"><span class="pre">-o</span> <span class="pre">riboprof_out/</span></code></div>
<div class="line">- All other arguments use the default value</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe trim -i riboprof_test/ -o riboprof_out/
</pre></div>
</div>
<div class="line-block">
<div class="line"><strong>Example 2 – Predict adapter and trim ribosome profiling (or single-end) sequence data:</strong></div>
<div class="line">- A minimum read length of 22 nucleotides after trimming is required in order to keep the read</div>
<div class="line">- A maximum or 6 processors can be used for the task</div>
<div class="line">- The <code class="xref py py-data docutils literal notranslate"><span class="pre">--adapters</span></code> argument was not passed, so an attempt to discover adapter sequences will be made (this is not always the most efficient or thorough method of trimming and providing the adapter sequences is recommended)</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe trim -i riboprof_test/ -o riboprof_out/ --min_length <span class="m">22</span> -m <span class="m">6</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><strong>Example 3 – Pass explicit adapter trim ribosome profiling (or single-end) sequence data:</strong></div>
<div class="line">- The default minimum read length threshold will be used</div>
<div class="line">- The maximum number of processors will be used by default</div>
<div class="line">- The <code class="xref py py-data docutils literal notranslate"><span class="pre">--adapters</span></code> argument was passed, so adapter sequences will trimmed explicitly</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe trim -i riboprof_test/ -o riboprof_out/ -a CTGTAGGCACCATCAAT
</pre></div>
</div>
<div class="line-block">
<div class="line"><strong>Example 4 – Predict adapter and trim paired-end sequence data:</strong></div>
<div class="line">- The <code class="xref py py-data docutils literal notranslate"><span class="pre">--adapters</span></code> argument was passed as <code class="xref py py-data docutils literal notranslate"><span class="pre">None</span> <span class="pre">None</span></code>, so an attempt to discover adapter sequences will be made for paired-end reads. The <code class="xref py py-data docutils literal notranslate"><span class="pre">-a</span> <span class="pre">None</span> <span class="pre">None</span></code> syntax is essential for <code class="xref py py-data docutils literal notranslate"><span class="pre">trim</span></code> to recognize the reads as paired-end</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe trim -i pe_test/ -o pe_out/ -a None None
</pre></div>
</div>
<div class="line-block">
<div class="line"><strong>Example 5 – Pass explicit adapter and trim paired-end sequence data:</strong></div>
<div class="line">- The <code class="xref py py-data docutils literal notranslate"><span class="pre">--adapters</span></code> argument was passed, so adapter sequences will trimmed explicitly</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe trim -i pe_test/ -o pe_out/ -a ACACTCTTTCCCTACACGACGCTCTTCCGATC GATCGGAAGAGCGGTTCAGCAGGAATGCCGAG
</pre></div>
</div>
<div class="line-block">
<div class="line"><strong>Example 6 – Trim single-end sequence data of polyX adapters:</strong></div>
<div class="line">- The <code class="xref py py-data docutils literal notranslate"><span class="pre">--adapters</span> <span class="pre">POLYX</span></code> argument was passed, so adapter sequences will trimmed of polyX sequences</div>
</div>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ xpresspipe trim -i se_test/ -o se_out/ -a POLYX
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="align.html" class="btn btn-neutral float-right" title="Alignment" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="analysis.html" class="btn btn-neutral float-left" title="Analysis" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Jordan A. Berg

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>